### PR TITLE
Performance: Getting rid of `array_merge` in loop

### PR DESCRIPTION
**Pull request by lbajsarowicz**
### Description (*)
**Work In Progress** Related with removing `foreach() { $x = array_merge($x, $this->x()); }`

### Fixed Issues (if relevant)
Nope

### Manual testing scenarios (*)
Nope

### Questions or comments
Nope.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Deploy/Package/Processor/PreProcessor/Css.php [hunk 1]
```diff
@@ -180,11 +180,11 @@ private function buildMap($packagePath, $filePath, $fullPath)
      */
     private function collectFileMap($fileName)
     {
-        $result = isset($this->map[$fileName]) ? $this->map[$fileName] : [];
+        $result = isset($this->map[$fileName]) ? [$this->map[$fileName]] : [[]];
```
Comment (TadeuRodrigues) [inline app/code/Magento/Deploy/Package/Processor/PreProcessor/Css.php]: can I propose you to use Elvis operator? like this 
$result = [$this->map[$fileName]] ?? [[]];

File: app/code/Magento/Tax/Model/ResourceModel/Calculation/Rate/Collection.php [hunk 2]
```diff
@@ -17,27 +29,27 @@ class Collection extends \Magento\Framework\Model\ResourceModel\Db\Collection\Ab
...
+        EntityFactory $entityFactory,
+        LoggerInterface $logger,
+        FetchStrategyInterface $fetchStrategy,
+        ManagerInterface $eventManager,
+        StoreManagerInterface $storeManager,
+        AdapterInterface $connection = null,
+        AbstractDb $resource = null
     ) {
         $this->_storeManager = $storeManager;
         parent::__construct($entityFactory, $logger, $fetchStrategy, $eventManager, $connection, $resource);
```
Comment (TadeuRodrigues) [inline app/code/Magento/Tax/Model/ResourceModel/Calculation/Rate/Collection.php:59]: maybe break in lines will be better to read.

**Review by TadeuRodrigues — CHANGES_REQUESTED** [review]

File: app/code/Magento/CatalogInventory/Model/StockIndex.php [hunk 3]
```diff
@@ -217,17 +227,19 @@ protected function getWebsitesWithDefaultStores($websiteId = null)
...
             /* @var $typeInstance AbstractType */
-            $parentIds = array_merge($parentIds, $typeInstance->getParentIdsByChild($productId));
+            $parentIds[] = $typeInstance->getParentIdsByChild($productId);
         }
 
-        if (!$parentIds) {
+        $parentIds = array_merge(...$parentIds);
+
+        if (empty($parentIds)) {
             return $this;
```
Comment (ihor-sviziev) [inline app/code/Magento/CatalogInventory/Model/StockIndex.php]: I think it should be like this to be valid
```suggestion
            return;
```

File: app/code/Magento/Tax/Model/ResourceModel/Calculation/Rate/Collection.php (hunk 2 above)
Comment (lbajsarowicz) [reply app/code/Magento/Tax/Model/ResourceModel/Calculation/Rate/Collection.php:59]: It's just passing by and line width is < 80 chars.

File: app/code/Magento/CatalogInventory/Model/StockIndex.php (hunk 3 above)
Comment (lbajsarowicz) [reply app/code/Magento/CatalogInventory/Model/StockIndex.php]: Nice catch!

File: app/code/Magento/ImportExport/Model/Import/ErrorProcessing/ProcessingErrorAggregator.php [hunk 4]
```diff
@@ -237,16 +237,12 @@ public function hasFatalExceptions()
...
-            return $result;
-        }
-
-        foreach (array_values($this->items['rows']) as $errors) {
-            $result = array_merge($result, $errors);
+            return [];
         }
 
-        return $result;
+        $errors = array_values($this->items['rows']);
```
Comment (ihor-sviziev) [inline app/code/Magento/ImportExport/Model/Import/ErrorProcessing/ProcessingErrorAggregator.php:244]: Looks like `$this->items['rows']` might be empty array, and we'll have error during array_merge

File: app/code/Magento/CatalogWidget/Block/Product/ProductsList.php [hunk 5]
```diff
@@ -173,7 +185,8 @@ protected function _construct()
 
         $this->addData([
             'cache_lifetime' => 86400,
-            'cache_tags' => [\Magento\Catalog\Model\Product::CACHE_TAG,
+            'cache_tags' => [
+                Product::CACHE_TAG,
             ], ]);
```
Comment (ihor-sviziev) [inline app/code/Magento/CatalogWidget/Block/Product/ProductsList.php]:
```suggestion
            ],
        ]);
```

File: app/code/Magento/PageCache/Model/Layout/LayoutPlugin.php [hunk 6]
```diff
@@ -80,10 +80,10 @@ public function afterGetOutput(\Magento\Framework\View\Layout $subject, $result)
                         continue;
                     }
                     // phpcs:ignore
```
Comment (ihor-sviziev) [inline app/code/Magento/PageCache/Model/Layout/LayoutPlugin.php]: I think this ignore could be removed

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/ImportExport/Model/Import/ErrorProcessing/ProcessingErrorAggregator.php (hunk 4 above)
Comment (lbajsarowicz) [reply app/code/Magento/ImportExport/Model/Import/ErrorProcessing/ProcessingErrorAggregator.php:244]: :+1:

File: app/code/Magento/PageCache/Model/Layout/LayoutPlugin.php (hunk 6 above)
Comment (lbajsarowicz) [reply app/code/Magento/PageCache/Model/Layout/LayoutPlugin.php]: :+1:

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-6595](https://jira.corp.magento.com/browse/ENGCOM-6595) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

File: app/code/Magento/Catalog/Block/Product/ProductList/Upsell.php [hunk 7]
```diff
@@ -262,11 +267,10 @@ public function getItemLimit($type = '')
...
-        foreach ($this->getItems() as $item) {
-            // phpcs:ignore Magento2.Performance.ForeachArrayMerge
-            $identities = array_merge($identities, $item->getIdentities());
-        }
-        return $identities;
+        $identities = array_map(function (DataObject $item) {
+            return $item->getIdentities();
+        }, $this->getItems());
+
+        return array_merge(...$identities);
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Block/Product/ProductList/Upsell.php:274]: I think there might be an issue when `$this->getItems()` will return empty array

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Block/Product/ProductList/Upsell.php (hunk 7 above)
Comment (lbajsarowicz) [reply app/code/Magento/Catalog/Block/Product/ProductList/Upsell.php:274]: Let me cover that case with unit tests

Comment (lbajsarowicz) [reply app/code/Magento/Catalog/Block/Product/ProductList/Upsell.php:274]: I wanted to write Unit Tests the right way, but I'll use the existing way to mock `items` - sooner or later, both blocks have to be rewritten testable-way.

Comment (lbajsarowicz) [reply app/code/Magento/Catalog/Block/Product/ProductList/Upsell.php:274]: You were right but not about `getItems`. Unit test caught it and I fixed this with fallback to `[[]]`

**Review by ihor-sviziev — APPROVED** [review]
Approved.

About failing tests:
Functional Tests B2B - looks not related to this PR
Semantic Version Checker - looks false positive

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-6595](https://jira.corp.magento.com/browse/ENGCOM-6595) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Comment by engcom-Golf** [comment]
SVC checker failure is not valid as method return type was not changed
DeepinScreenshot_select-area_20200124105550
DeepinScreenshot_select-area_20200124110447

**Comment by lbajsarowicz** [comment]
@engcom-Golf I rolled back change of return type to have 100% green.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
@lbajsarowicz thank you for reverting changes that require approval. Could you also resolve merge conflict that just appeared?

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-6595](https://jira.corp.magento.com/browse/ENGCOM-6595) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Review by TadeuRodrigues — APPROVED** [review]

**Pull request merged**